ambry-replication/src/main/java/com/github/ambry/replication/ReplicationManager.java

```diff
@@ -451,4 +487,197 @@ public void onPartitionBecomeDroppedFromOffline(String partitionName) {
+     * found during previous meta data exchange. If there are matching messages (based on store key), remove them from the missing message set.
+     * This is used during leader-based replication to update token for standby replicas. Standby replicas store the
+     * missing messages in metadata exchange, track them through intra-dc replication and update token when all the
+     * missing messages are written to store.
+     * @param partitionId partition ID of the messages written to store
+     * @param messageInfoList list of messages written to store
+     */
+    void onMessageWriteForPartition(PartitionId partitionId, List<MessageInfo> messageInfoList) {
+      rwLock.readLock().lock();
+      try {
+        PartitionInfo partitionInfo = partitionToPartitionInfo.get(partitionId);
+        partitionInfo.updateReplicaInfosOnMessageWrite(messageInfoList);
+      } finally {
+        rwLock.readLock().unlock();
+      }
+    }
+
+    /**
+     * Add a leader partition and its set of peer leader replicas.
+     * @param partitionName name of the partition to be added
+     */
+    public void addLeaderPartition(String partitionName) {
+      partitionLeaderInfo.addPartition(partitionName);
+    }
+
+    /**
+     * Remove a leader partition from the map of leader partitions.
+     * @param partitionName name of the partition to be removed
+     */
+    public void removeLeaderPartition(String partitionName) {
+      partitionLeaderInfo.removePartition(partitionName);
+    }
+
+    /**
+     * Refreshes the list of remote leaders for all leader partitions by querying the latest information from
+     * RoutingTableSnapshots of all data centers.
+     */
+    public void refreshPeerLeadersForLeaderPartitions() {
+      partitionLeaderInfo.refreshPeerLeadersForAllPartitions();
+    }
+
+    /**
+     * Get a map of partitions to their sets of peer leader replicas (this method is only by ReplicationTest for now)
+     * @return an unmodifiable map of peer leader replicas stored by partition
+     */
+    public Map<String, Set<ReplicaId>> getPeerLeaderReplicasByPartition() {
+      return partitionLeaderInfo.getPeerLeaderReplicasByPartition();
+    }
+
+    /**
+     * Checks if a remote replica is a leader for a partition (Pre-requisite: the partition itself should be a leader locally).
+     * @param partitionName name of local leader partition
+     * @param replicaId remote replica to be checked
+     * @return true if remote replica is a leader for a partition (Pre-requisite: the partition itself should be a leader locally).
+     */
+    public boolean isPeerReplicaLeaderForPartition(String partitionName, ReplicaId replicaId) {
+      return partitionLeaderInfo.isPeerReplicaLeaderForPartition(partitionName, replicaId);
+    }
+
+    /**
+     * Maintains the list of leader partitions on local node and their corresponding peer leaders in remote data centers
+     */
+    class PartitionLeaderInfo {
```

Comment:
Could PartitionLeaderInfo be compressed into LeaderBasedReplicationAdmin unless there is a future use case for PLI to be used separately from LBRA? It seems like most methods of LBRA just call the similarly named method of PLI.

Reply:
Sure, compressed PartitionLeaderInfo into LeaderBasedReplicationAdmin.